File: stepik-union/src/test/java/org/stepik/core/courseFormat/SerializationTest.java
package org.stepik.core.courseFormat;

import org.stepik.core.StepikProjectManager;
import com.thoughtworks.xstream.XStream;
import org.jetbrains.annotations.NotNull;
import org.junit.Test;
import org.stepik.api.objects.lessons.CompoundUnitLesson;
import org.stepik.api.objects.steps.Limit;
import org.stepik.api.objects.steps.Step;
import org.stepik.core.TestUtils;
import org.xml.sax.SAXException;

import javax.xml.parsers.ParserConfigurationException;
import java.io.IOException;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

/**
 * @author meanmail
 */
public class SerializationTest {

    @Test
    public void serializeCourseNode()
            throws IOException, SAXException, ParserConfigurationException, InstantiationException, IllegalAccessException {
        CourseNode node = new CourseNode();
        node.getData();
        serialize("CourseNode", node);
    }

    @Test
    public void serializeSectionNode()
            throws IOException, SAXException, ParserConfigurationException, InstantiationException, IllegalAccessException {
        SectionNode node = new SectionNode();
        node.getData();
        LessonNode lessonNode = new LessonNode();
        lessonNode.setParent(node);
        CompoundUnitLesson data = lessonNode.getData();
        assertNotNull(data);
        data.getLesson();
        data.getUnit();
        node.getChildren().add(lessonNode);
        serialize("SectionNode", node);
    }

    @Test
    public void serializeLessonNode()
            throws IOException, SAXException, ParserConfigurationException, InstantiationException, IllegalAccessException {
        LessonNode node = new LessonNode();
        CompoundUnitLesson data = node.getData();
        assertNotNull(data);
        data.getLesson();
        data.getUnit();
        serialize("LessonNode", node);
    }

    @Test
    public void serializeStepNode()
            throws IOException, SAXException, ParserConfigurationException, InstantiationException, IllegalAccessException {
        StepNode node = new StepNode();
        node.getData();
        node.setId(100);
        Limit limit = new Limit();
        limit.setMemory(256);
        limit.setTime(8);
        Step data = node.getData();
        assertNotNull(data);
        data.getBlock().getOptions().getLimits().put("Java 8", limit);
        serialize("StepNode", node);
    }

    private void serialize(@NotNull String name, @NotNull StudyNode node) throws IOException {
        XStream xs = StepikProjectManager.getXStream();

        String expected = TestUtils.readTextFile(SerializationTest.class, String.format("expected%s.xml", name));

        assertEquals(name, expected, xs.toXML(node));
    }
}

File: stepik-union/src/test/java/org/stepik/core/courseFormat/SerializationTest.kt
package org.stepik.core.courseFormat

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Test
import org.stepik.api.objects.steps.Limit
import org.stepik.core.StepikProjectManager
import org.stepik.core.TestUtils
import org.xml.sax.SAXException
import java.io.IOException
import javax.xml.parsers.ParserConfigurationException


class SerializationTest {

    @Test
    @Throws(IOException::class,
            SAXException::class,
            ParserConfigurationException::class,
            InstantiationException::class,
            IllegalAccessException::class)
    fun serializeCourseNode() {
        val node = CourseNode()
        node.data
        serialize("CourseNode", node)
    }

    @Test
    @Throws(IOException::class, SAXException::class,
            ParserConfigurationException::class,
            InstantiationException::class,
            IllegalAccessException::class)
    fun serializeSectionNode() {
        val node = SectionNode()
        node.data
        val lessonNode = LessonNode()
        lessonNode.parent = node
        val data = lessonNode.data
        assertNotNull(data)
        data!!.lesson
        data.unit
        node.children.add(lessonNode)
        serialize("SectionNode", node)
    }

    @Test
    @Throws(IOException::class, SAXException::class,
            ParserConfigurationException::class,
            InstantiationException::class,
            IllegalAccessException::class)
    fun serializeLessonNode() {
        val node = LessonNode()
        val data = node.data
        assertNotNull(data)
        data!!.lesson
        data.unit
        serialize("LessonNode", node)
    }

    @Test
    @Throws(IOException::class, SAXException::class,
            ParserConfigurationException::class,
            InstantiationException::class,
            IllegalAccessException::class)
    fun serializeStepNode() {
        val node = StepNode()
        node.data
        node.id = 100
        val limit = Limit()
        limit.memory = 256
        limit.time = 8
        val data = node.data
        assertNotNull(data)
        data!!.block.options.limits["Java 8"] = limit
        serialize("StepNode", node)
    }

    @Throws(IOException::class)
    private fun serialize(name: String, node: StudyNode<*, *>) {
        val xs = StepikProjectManager.xStream

        val expected = TestUtils.readTextFile(SerializationTest::class.java, String.format("expected%s.xml", name))

        assertEquals(name, expected, xs.toXML(node))
    }
}
